### PR TITLE
feat: add @SdkAdvancedApi annotation for error-prone APIs

### PR DESCRIPTION
**Pull request by zoewangg**
## Motivation and Context

The SDK exposes several extension points that customers implement, override, call, or configure to customize streaming and request/response processing. Each carries a strict, often subtle contract, and a non-compliant use compiles cleanly but fails or misbehaves at runtime rather than reporting a clear error. Today that risk is described only in prose Javadoc, written differently per API and in a form no tool can read. This change adds a single, uniform, machine-readable marker for that class of error-prone API, which is the prerequisite for future build-time or static-analysis/AI tooling that can warn on misuse.

## Modifications

- Add `@SdkAdvancedApi` in `core/annotations` (CLASS retention, `@Documented`, targets TYPE/METHOD/CONSTRUCTOR/FIELD). Members: `caution` (required; WHEN_IMPLEMENTED / WHEN_OVERRIDDEN / WHEN_CONFIGURED / WHEN_CALLED), `guidance` (required), `saferAlternative` (optional), `link` (optional).
- Apply it to the error-prone extension points: `AsyncRequestBody` (type + `fromPublisher`), `AsyncResponseTransformer` (type + `toPublisher()` / `toPublisher(Duration)`), `ResponseTransformer`, `ContentStreamProvider`, the mutating `ExecutionInterceptor` hooks (`modifyHttpContent`, `modifyAsyncHttpContent`), and the `FUTURE_COMPLETION_EXECUTOR` advanced client option.
- Add a checkstyle check capping `guidance`/`saferAlternative` length at 1000 characters.
- The annotation is advisory only: it does not gate compilation and adds no runtime behavior.

## Testing

- `mvn install` on `core/annotations`, `http-client-spi`, `core/sdk-core`, and `build-tools`: BUILD SUCCESS.
- japicmp reports no binary/source-incompatible modifications in any affected module (adding a CLASS-retention marker is compatible).
- New checkstyle check has 5 unit tests (pass/fail/concatenation/boundary cases).

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added a changelog entry (scripts/new-change)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/async/AsyncResponseTransformer.java [hunk 1]
```diff
@@ -86,6 +88,16 @@
...
 @SdkPublicApi
+@SdkAdvancedApi(
+    caution = Caution.WHEN_IMPLEMENTED,
+    guidance = "prepare() is called on each request attempt; if the CompletableFuture it returned on a previous "
+          + "attempt has already completed, it must return a new instance so the result of a retry is not lost, and "
+          + "exceptionOccurred() should release any resources the attempt opened (for example an open file channel) "
+          + "so retries do not leak them. onStream() receives the response body as a reactive-streams Publisher that "
+          + "the implementation subscribes to, and that subscriber must comply with the reactive-streams "
+          + "specification; a subscriber that never "
+          + "requests data stalls the response.",
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/async/AsyncResponseTransformer.java]: nit: fold into the previous line

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/SdkAdvancedAsyncClientOption.java [hunk 2]
```diff
@@ -52,6 +54,13 @@ public final class SdkAdvancedAsyncClientOption<T> extends ClientOption<T> {
      * within the I/O thread because it may block the I/O thread and cause deadlock, especially if you are sending
      * another SDK request in the {@link CompletableFuture} chain since the SDK may perform blocking calls in some cases.
      */
+    @SdkAdvancedApi(
+        caution = Caution.WHEN_CONFIGURED,
+        guidance = "An executor that runs on the client's I/O threads (e.g. Runnable::run) can block "
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/SdkAdvancedAsyncClientOption.java]: maybe remove "client's I/O thread" and just say "that does not dispatch to a separate thread"

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/sync/ResponseTransformer.java [hunk 3]
```diff
@@ -75,6 +77,12 @@
  */
 @FunctionalInterface
 @SdkPublicApi
+@SdkAdvancedApi(
+    caution = Caution.WHEN_IMPLEMENTED,
+    guidance = "transform() must fully drain and close the response stream and honor thread interrupts. A "
```
Comment (Fred1155) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/sync/ResponseTransformer.java]: This seems contradict with Javadoc, and is the "must close" wording accurate? I think it does not need to close the response stream since CombinedResponseHandler will close it?

Comment (zoewangg) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/sync/ResponseTransformer.java]: Good catch! I don't think we actually need advanced API annotation for this API, will remove

**Review by Fred1155 — APPROVED** [review]

**Review by dagnir — APPROVED** [review]
